### ISSUE TITLE
BUG: child keyed elements of a component are not consistently mapped to the proper Component context

### ISSUE THREAD

**Issue by jasonmacdonald** [marko-js/marko#1009]
### Marko Version: x.x.x
4.7.5

### Details

**UPDATE: After further investigation, I'm not convinced this is related to what I said below, but I'll leave it here.**

@DylanPiercey THe commit you made [here](https://github.com/marko-js/marko/commit/aed62f503f35c3e572b53240843786dfd8d45715#diff-66b7557cb75005002decad6a4b3db15f) has caused a new bug. Basically, there are scenarios where the comment tags Marko puts around components to identify their __component have their endNode removed, but the start node remains, cause elements to get keyed into the wrong parent component. When this happens, if the node is removed the key is never cleaned up, because the clean-up process is checking the wrong component.___keyedElements[key] for the key. It's really hard to explain, and I've spent WAY too much time hunting this down, but the end effect is that if you listen for attach/detach events, say for transitions, certain elements will never fire an attach event the second time they are added because the key still exists. The example below may explain the behavior better than I can. 

From what I can tell, this is caused by the extra comment element that is added by morphdom from the above PR. If  `var beforeChild = startNode.previousSibling;` and  `var afterChild = endNode.nextSibling;` are both null, and `startNode` and `endNode` are a pair of `<!-- $marko -->` elements wrapping a component, when the new `<!-- $marko -->` comment is added for `tempChild` but the `afterChild` is left blank, the final output is missing the closing component `<!-- marko -->`. This makes the child element of that component get keyed to the parent element, instead of child... or something. Frankly, it's really confusing and I'm probably doing a terrible job of explaining it so I'll just leave the test case here. :)

## Main Component
```marko
class {
    onInput(input) {
        this.state = Object.assign({index: 1}, input);
    }

    handleClick(index) {
        this.state.index = this.state.index === 1 ? 2 : 1;
    }

    onAttach(name, event, el){
        if(name == 'div1'){
            name += ' +++ Only see this once +++';
        }
        console.log('attached', name);
    }

    onDetach(name, event, el){
        console.log('detached', name);
    }

}

<transition>
    <div if(state.index === 1)  onAttach('onAttach', 'div1') onDetach('onDetach', 'div1')>test: ${state.index}</div>
    <div else-if(state.index === 2) onAttach('onAttach', 'div2') onDetach('onDetach', 'div2')>test: ${state.index}</div>
</transition>


<br />
<button onClick('handleClick')>
    Click Me
</button>

```

## Trainsition Component
```marko
class {

}
<main>
    <include(input.renderBody) />
</main>
```
### Expected Behavior
When clicking the button, you should see the `attach` and `detach` events for each div as they switch back and forth in the console.

### Actual Behavior
The div1 only ever fires `attach` on the initial render. After that, only the detach event is fired for div1, yet div2 always fires.

### Possible Fix
One way to get around this and have everything work is to simply add another element __after__ the 2 divs.

```marko
<transition>
    <div if(state.index === 1)  onAttach('onAttach', 'div1') onDetach('onDetach', 'div1')>test: ${state.index}</div>
    <div else-if(state.index === 2) onAttach('onAttach', 'div2') onDetach('onDetach', 'div2')>test: ${state.index}</div>
    <br/>
</transition>
```

This seems to allow the `afterChild` to not be null, and thereby not delete the `<!-- $marko -->` end node (I think), therefore both divs get their keys mapped to the proper component. Obviously, though, this is not the correct solution

### Is this something you're interested in working on?
I really want to provide a fix, but I assume reverting that PR would just bring back the old server-side render issue. And what I have tried so far has been very finicky, but I "think" simply adding a temp end node would solve this issue, but it's possible you just wind up with the same bug from the original issue. If you could provide a little guidance, I'd be happy to do the actual fix and add some tests.

**Comment by jasonmacdonald** [comment]
Up, I tried adding an additional comment after to see if that worked, but it had no effect. So clearly there is something else going on, which causes the first div's key to be mapped to `b1`, but the second div's key to be mapped to `b1-0` (the main tag), which is correct (I believe).

**Comment by jasonmacdonald** [comment]
Ok, I think I found the solution. The problem is when morphdom is looping through childNodes and makes it all the way to the end of the while loop to detach any lingering nodes. IT passes `component` from the outer most morphChildren call, but the element being removed is actually keyed on the parent component (which makes sense when you think about how transclusion works, the `divs` are mounted to the main component instance, not the transition component, which is what is being checked against with `component`. Now, I assume that when you add an element on the end, morphdom has to move that last element into the "second" position using insertAfter, which triggers the if statement `if(curFromNodeChild)` to be true, which calls `detachNode()` using the component context of `componentFromNode`(main component ID: b1), instead of the outer `component`(transitions component ID: b1-0 ), which causes it to properly detach the key. Without that, there is no "moving" of elements so it cleanly exits the while loop and enters the final while loop, which now tries to detach the element using `component` (b1-0) instead of `componentForNode` (b1).

Now, there might be a better way to address this, but I was able to fix it by adding the following check to the `destroyNodesRecursively` call...

```javascript
       if(node.___markoVElement && node.___markoVElement.___component !== component){
            // use the nodes linked component
            component = node.___markoVElement.___component;
        }
```

making it

```javascript
function destroyNodeRecursive(node, component) {
    if (node.nodeType === 1) {
        var key;
        
        // ###### NEW CODE HERE #######
        if(node.___markoVElement && node.___markoVElement.___component !== component){
            // use the nodes linked component
            component = node.___markoVElement.___component;
        }
        // ###### END ########

        if (component && (key = node.___markoKey)) {
            if (node === component.___keyedElements[key]) {
                delete component.___keyedElements[key];
            }
        }

        var curChild = node.firstChild;
        while(curChild) {
            destroyComponentForNode(curChild);
            destroyNodeRecursive(curChild, component);
            curChild = curChild.nextSibling;
        }
    }
}
```

I ran the test suite after this change and everything appears to still pass. Should I submit a PR or do you guys want to tackle this differently?

**Comment by mlrawlings** [comment]
Closed via #1013 and published in `4.9.1`

**Comment by jasonmacdonald** [comment]
Thanks, Michael!!
